src/interpreter/balances.rs

```diff
@@ -0,0 +1,191 @@
+    {
+        iter.into_iter()
+            .sorted_by_key(|k| k.0)
+            .enumerate()
+            .try_fold(HashMap::new(), |mut state, (i, (asset, balance))| {
+                let offset = VM_MEMORY_BALANCES_OFFSET + i * (AssetId::LEN + WORD_SIZE);
+
+                state
+                    .entry(asset)
+                    .or_insert(Balance::new(0, offset))
+                    .checked_add(balance)
+                    .ok_or(ValidationError::ArithmeticOverflow)?;
+
+                Ok(state)
+            })
+            .map(|state| Self { state })
+    }
+
+    /// Fetch the balance of a given Id, if set.
+    pub fn balance(&self, asset: &AssetId) -> Option<Word> {
+        self.state.get(asset).map(Balance::value)
+    }
+
+    fn _set_memory_balance(balance: &Balance, memory: &mut [u8]) -> Word {
+        let value = balance.value();
+        let offset = balance.offset();
+
+        let offset = offset + AssetId::LEN;
+        let memory = &mut memory[offset..offset + WORD_SIZE];
+
+        memory.copy_from_slice(&value.to_be_bytes());
+
+        value
+    }
+
+    /// Attempt to add the balance of an asset, updating the VM memory in the appropriate
+    /// offset
+    pub fn checked_balance_add(&mut self, memory: &mut [u8], asset: &AssetId, value: Word) -> Option<Word> {
+        self.state
+            .get_mut(asset)
+            .and_then(|b| b.checked_sub(value))
+            .map(|balance| Self::_set_memory_balance(balance, memory))
+            .or((value == 0).then(|| 0))
+    }
+
+    /// Attempt to subtract the balance of an asset, updating the VM memory in the appropriate
+    /// offset
+    pub fn checked_balance_sub(&mut self, memory: &mut [u8], asset: &AssetId, value: Word) -> Option<Word> {
+        self.state
+            .get_mut(asset)
+            .and_then(|b| b.checked_sub(value))
+            .map(|balance| Self::_set_memory_balance(balance, memory))
+            .or((value == 0).then(|| 0))
+    }
+
+    /// Write all assets into the VM memory.
+    pub fn to_vm<S>(self, vm: &mut Interpreter<S>) {
+        let len = vm.params().max_inputs * (AssetId::LEN + WORD_SIZE) as Word;
+
+        vm.registers[REG_SP] += len;
+        vm.reserve_stack(len)
+            .expect("consensus parameters won't allow stack overflow for VM initialization");
+
+        self.state.iter().for_each(|(asset, balance)| {
+            let value = balance.value();
+            let ofs = balance.offset();
+
+            (&mut vm.memory[ofs..ofs + AssetId::LEN]).copy_from_slice(asset.as_ref());
+            (&mut vm.memory[ofs + AssetId::LEN..ofs + AssetId::LEN + WORD_SIZE]).copy_from_slice(&value.to_be_bytes());
+        });
+
+        vm.balances = self;
+    }
+}
+
+impl Index<&AssetId> for RuntimeBalances {
+    type Output = Word;
+
+    fn index(&self, index: &AssetId) -> &Self::Output {
+        &self.state[index].value
+    }
+}
+
+#[test]
+fn runtime_balances_writes_to_memory_correctly() {
```

Comment:
we should also have unit tests that verify checked_add and checked_sub update the in-memory values properly.

Reply:
Agreed, done here

https://github.com/FuelLabs/fuel-vm/pull/173/commits/902e894408a347987f157371eea947322ecb6352